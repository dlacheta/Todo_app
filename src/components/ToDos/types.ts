export type FormDataType = {
  todo: string;
};
