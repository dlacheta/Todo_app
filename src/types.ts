export type TodosType = {
  id: string;
  todo: string;
  completed: boolean;
};

export enum FilterEnum {
  all,
  active,
  completed,
}
