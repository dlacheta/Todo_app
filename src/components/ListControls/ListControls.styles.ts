import tw from 'twin.macro';
import styled from 'styled-components';

export const TodosListInfo = tw.div`flex flex-row border-y-2 border-solid border-slate-200 text-sm justify-between`;

export const TodosListItemsLeft = tw.div`p-4`;

export const TodosListControls = tw.div`flex flex-row justify-around`;

export const TodosListControl = styled.div<{
  isActive?: boolean;
  isDisabled?: boolean;
}>`
  ${tw`p-4 cursor-pointer`}
  ${(props) => props.isActive && tw`bg-slate-200 text-red-800`}
  ${(props) => props.isDisabled && tw`text-slate-200 cursor-default`}
`;
