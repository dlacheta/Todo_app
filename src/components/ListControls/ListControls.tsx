import React from 'react';
import {
  TodosListInfo,
  TodosListControls,
  TodosListControl,
  TodosListItemsLeft,
} from './ListControls.styles';
import { FilterEnum } from '../../types';

type Props = {
  todosLeft: number;
  isTodosCompleted: boolean;
  filter: FilterEnum;
  onFilterChange: (filter: FilterEnum) => void;
  onClearCompleted: () => void;
};

const ListControls = ({
  todosLeft,
  isTodosCompleted,
  filter,
  onFilterChange,
  onClearCompleted,
}: Props) => {
  return (
    <TodosListInfo>
      <TodosListItemsLeft>{todosLeft} left</TodosListItemsLeft>
      <TodosListControls>
        <TodosListControl
          data-testid="all-control"
          isActive={filter === FilterEnum.all}
          onClick={() => {
            onFilterChange(FilterEnum.all);
          }}
        >
          All
        </TodosListControl>
        <TodosListControl
          data-testid="active-control"
          isActive={filter === FilterEnum.active}
          onClick={() => {
            onFilterChange(FilterEnum.active);
          }}
        >
          Active
        </TodosListControl>
        <TodosListControl
          data-testid="completed-control"
          isActive={filter === FilterEnum.completed}
          onClick={() => {
            onFilterChange(FilterEnum.completed);
          }}
        >
          Completed
        </TodosListControl>
      </TodosListControls>
      <TodosListControl
        data-testid="clear-completed-control"
        isDisabled={!isTodosCompleted}
        onClick={() => {
          isTodosCompleted && onClearCompleted();
        }}
      >
        Clear completed
      </TodosListControl>
    </TodosListInfo>
  );
};

export default ListControls;
