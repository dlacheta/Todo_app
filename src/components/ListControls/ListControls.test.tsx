import React from 'react';
import { render as rtlRender, fireEvent } from '@testing-library/react';
import ListControls from './ListControls';
import { FilterEnum } from '../../types';

const render = (newProps?: any) => {
  const props = {
    todosLeft: 0,
    isTodosCompleted: true,
    filter: FilterEnum.all,
    onFilterChange: jest.fn(),
    onClearCompleted: jest.fn(),
    ...newProps,
  };
  const wrapper = rtlRender(<ListControls {...props} />);

  return {
    wrapper,
    props,
  };
};

test('should render component', () => {
  const {
    wrapper: { container },
  } = render();

  expect(container).toMatchSnapshot();
});

test('should run onFilterChange() with all', () => {
  const {
    wrapper: { getByTestId },
    props,
  } = render();

  const container = getByTestId('all-control');
  fireEvent.click(container);

  expect(props.onFilterChange).toHaveBeenCalledWith(FilterEnum.all);
});

test('should run onFilterChange() with active', () => {
  const {
    wrapper: { getByTestId },
    props,
  } = render();

  const container = getByTestId('active-control');
  fireEvent.click(container);

  expect(props.onFilterChange).toHaveBeenCalledWith(FilterEnum.active);
});

test('should run onFilterChange() with completed', () => {
  const {
    wrapper: { getByTestId },
    props,
  } = render();

  const container = getByTestId('completed-control');
  fireEvent.click(container);

  expect(props.onFilterChange).toHaveBeenCalledWith(FilterEnum.completed);
});

test('should run onClearCompleted()', () => {
  const {
    wrapper: { getByTestId },
    props,
  } = render();

  const container = getByTestId('clear-completed-control');
  fireEvent.click(container);

  expect(props.onClearCompleted).toHaveBeenCalled();
});
