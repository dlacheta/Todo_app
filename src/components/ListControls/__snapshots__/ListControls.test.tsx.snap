// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<div>
  <div
    class="ListControlsstyles__TodosListInfo-sc-8el6kk-0 iSAPco"
  >
    <div
      class="ListControlsstyles__TodosListItemsLeft-sc-8el6kk-1 erjySO"
    >
      0
       left
    </div>
    <div
      class="ListControlsstyles__TodosListControls-sc-8el6kk-2 kPhJAB"
    >
      <div
        class="ListControlsstyles__TodosListControl-sc-8el6kk-3 giciPs"
        data-testid="all-control"
      >
        All
      </div>
      <div
        class="ListControlsstyles__TodosListControl-sc-8el6kk-3 eEyFCb"
        data-testid="active-control"
      >
        Active
      </div>
      <div
        class="ListControlsstyles__TodosListControl-sc-8el6kk-3 eEyFCb"
        data-testid="completed-control"
      >
        Completed
      </div>
    </div>
    <div
      class="ListControlsstyles__TodosListControl-sc-8el6kk-3 eEyFCb"
      data-testid="clear-completed-control"
    >
      Clear completed
    </div>
  </div>
</div>
`;
